tracee-rules/signatures/golang/stdio_over_socket.go

```diff
@@ -0,0 +1,159 @@
+package main
+
+import (
+	"encoding/json"
+	"fmt"
+	"github.com/aquasecurity/tracee/tracee-rules/types"
+	tracee "github.com/aquasecurity/tracee/tracee/external"
+	"strings"
+)
+
+type connectAddrData struct {
+	SaFamily string `json:"sa_family"`
+	SinPort  string `json:"sin_port"`
+	SinAddr  string `json:"sin_addr"`
+}
+
+type stdIoOverSocket struct {
+	cb              types.SignatureHandler
+	processSocketIp map[int]map[int]string
+}
+
+func (sig *stdIoOverSocket) Init(cb types.SignatureHandler) error {
+	sig.cb = cb
+	sig.processSocketIp = make(map[int]map[int]string)
+
+	return nil
+}
+
+func (sig *stdIoOverSocket) GetMetadata() (types.SignatureMetadata, error) {
+	return types.SignatureMetadata{
+		Name:        "STD I/O Over Socket",
+		Description: "Redirection of process's standard input/output to socket",
+		Tags:        []string{"linux", "container"},
+		Properties: map[string]interface{}{
+			"Severity":     3,
+			"MITRE ATT&CK": "Persistence: Server Software Component",
+		},
+	}, nil
+}
+
+func (sig *stdIoOverSocket) GetSelectedEvents() ([]types.SignatureEventSelector, error) {
+	return []types.SignatureEventSelector{
+		{Source: "tracee", Name: "connect"},
+		{Source: "tracee", Name: "dup2"},
+		{Source: "tracee", Name: "close"},
+	}, nil
+}
+
+func (sig *stdIoOverSocket) OnEvent(e types.Event) error {
+
+	eventObj, ok := e.(tracee.Event)
+	if !ok {
+		return fmt.Errorf("invalid event")
+	}
+
+	var connectData connectAddrData
+	socketIpMap := make(map[int]string)
+	stdAll := []int{0, 1, 2}
+
+	pid := eventObj.ProcessID
+
+	switch eventObj.EventName {
+
+	case "connect":
+
+		sockfdArg, err := GetTraceeArgumentByName(eventObj, "sockfd")
+		if err != nil {
+			return err
+		}
+
+		sockfd := sockfdArg.Value.(int)
+
+		addrArg, err := GetTraceeArgumentByName(eventObj, "addr")
+		if err != nil {
+			return err
+		}
+
+		addrStr := strings.Replace(addrArg.Value.(string), "'", "\"", -1)
+		err = json.Unmarshal([]byte(addrStr), &connectData)
+		if err != nil {
+			return fmt.Errorf(err.Error())
+		}
+
+		if connectData.SaFamily == "AF_INET" {
+			socketIpMap[sockfd] = connectData.SinAddr
+			sig.processSocketIp[pid] = socketIpMap
+		}
+
+	case "dup2":
+
+		_, pidExists := sig.processSocketIp[pid]
+
+		if pidExists {
+
+			oldFdArg, err := GetTraceeArgumentByName(eventObj, "oldfd")
+			if err != nil {
+				return err
+			}
+
+			srcFd := oldFdArg.Value.(int)
+
+			newFdArg, err := GetTraceeArgumentByName(eventObj, "newfd")
+			if err != nil {
+				return err
+			}
+
+			dstFd := newFdArg.Value.(int)
+
+			ip, socketfdExists := sig.processSocketIp[pid][srcFd]
+
+			// this means that a socket FD is duplicated into one of the standard FDs
+			if socketfdExists && intInSlice(dstFd, stdAll) && !intInSlice(srcFd, stdAll) {
```

Comment:
is it possible that `intInSlice(srcFd, stdAll)`? to my understanding it isn't because it's the socket fd, in this case the final predicate is unneeded

Reply:
you're right. resolved.